### PR TITLE
Fix NSTableView animations

### PR DESCRIPTION
**Pull request by hannesoid**
# NSTableView
- Now based on patches, because NSTableView needs moves to be executed in order
- Changed Interface to use rowIndices, because there are not IndexPaths involved in the row animations of it
- I deprecated the previous methods for compatibility

# UITableView
- ~~prevent beginUpdates/endUpdates for empty diff~~ removed after discussion

### CODE REVIEW

**Comment by jenox** [comment]
I don't think we should drop the calls to `beginUpdates()` and `endUpdates()` if the diff is empty. `beginUpdates()` and `endUpdates()` also give the visible cells a chance to resize with animation, which users of this library might rely on. Keep in mind that an empty diff does not imply equal or substitutable data: Diffing is generally done based on record identity (Identifiable), and not based on substitutability (Equatable).

For example, an item in your shopping cart might change its price. Differ would leave this cell as is when diffing based on the items' IDs, but the cell still needs to redraw and potentially requires more space to do so.

**Comment by hannesoid** [comment]
Thanks for the feedback, good point. For my use case I wanted it to exactly prevent the behaviour because I handled those needed resizes/redraws more specifically. But I understand now that people may rely on it and expect it. Would it be alright to add a parameter defaulting to the always update behaviour like: updateForEmptyDiff: Bool = true or preventAnimationForEmptyDiff: Bool = false, or would you say in this case the callsite should be responsible for it (first creating diff then checking if empty and then calling or not?)?

**Comment by jenox** [comment]
Intuitively, I'd want to avoid a `preventAnimationForEmptyDiff` parameter. However, moving this responsibility to the call site isn't nice either, as we'd end up diffing twice.

Could you elaborate on what you are currently doing to handle the resizes for specifically?

**Comment by hannesoid** [comment]
I have a UITableView and NSTableView with "self-sizing" cells that involve text editing which loops through a model layer. In order to prevent loss of firstResponder, other potential glitches and for keeping the updates to the minimum necessary, I do the following when a new list of items "arrives": For items which have been in the old list as well and have changed (based on full equality check), and have a visible cell, I update their contentView directly and take note of the potential changed height, on iOS by wrapping in begin/endUpdates and on macOS by calling `noteHeightOfRows(withIndexesChanged:)`. 
After that I do an animated update using Differ with equality check being based on identifiers.

**Comment by jenox** [comment]
Personally I don't see a reason to use separate `beginUpdates()`/`endUpdates()` pairs for the record identity and equality checks. But then again I haven't worked much with `firstResponder`s in dynamic `UITableView`s.

[Here](https://github.com/Chrono24/Swift-Value-Types/blob/master/Sources/ValueTypeExample.swift#L62)'s how I usually update my table views. Maybe you can give that a try? In your case with dynamic row heights, you'd want to run the block to update the cell's models _before_ calling out to Differ, such that they are configured correctly by the time `endUpdates()` is called and can compute the correct row height. Note that calling out to Differ from a `didSet` as in the linked project is generally a bad idea (#62), so you'd probably want something like this instead:

```swift
func transition(to newItems: [Item]) {
    for indexPath in self.tableView.indexPathsForVisibleRows ?? [] {
        if let cell = self.tableView.cellForRow(at: indexPath) as? YourCellType {
            let newItem = newItems[indexPath.row]

            if self.items.containsDifferentSnapshot(of: newItem) {
                cell.item = newItem
            }
        }
    }

    self.tableView.animateRowChanges(
        oldData: self.items,
        newData: newItems,
        isEqual: { $0.id == $1.id },
        updateData: { self.items = newItems } // See #62, coming soon™
    )
}
```

```swift
extension Collection where Element: Equatable & Identifiable {
    public func containsDifferentSnapshot(of element: Element) -> Bool {
        return contains(where: { $0.id == element.id && $0 != element })
    }
}
```

Please let me know if this works for you.

**Comment by hannesoid** [comment]
I do a similar approach in my project, but can't go into more detail now.
But as for this PR, I think I'll just leave out that new option parameter and use the current behavior, the scope then being just to fix `NSTableView` updates.

**Comment by hannesoid** [comment]
@tonyarnold fyi this is a the new PR after discarding #69

File: Sources/Differ/Diff+AppKit.swift [hunk 1]
```diff
@@ -10,38 +10,92 @@ extension NSTableView {
...
+        isEqual: EqualityChecker<T>,
+        deletionAnimation: NSTableView.AnimationOptions = [],
+        insertionAnimation: NSTableView.AnimationOptions = [],
+        rowIndexTransform: (Int) -> Int = { $0 }
+    ) {
+        withoutActuallyEscaping(isEqual) { isEqual in
+            func wrap(_ wrapped: T.Element) -> CustomEquatable<T> { CustomEquatable(isEqual: isEqual, wrapped: wrapped) }
+
+            animateRowChanges(oldData: oldData.map(wrap), newData: newData.map(wrap), deletionAnimation: deletionAnimation, insertionAnimation: insertionAnimation, rowIndexTransform: rowIndexTransform)
+        }
```
Comment (jenox) [inline Sources/Differ/Diff+AppKit.swift:46]: I don't like calling `animateRowChanges(oldData: newData:deletionAnimation:insertionAnimation:rowIndexTransform:)` from here, as it forces us to create this `Equatable` wrapper and work with `withoutActuallyEscaping(_:)`.

Can we maybe turn this around, i.e. compute the patch based on `isEqual` and apply it in here, and call this method with `isEqual: ==` from `animateRowChanges(oldData: newData:deletionAnimation:insertionAnimation:rowIndexTransform:)`? This would require adding an `isEqual:` parameter to the patch-family of methods, which might be helpful to other applications as well.

File: Sources/Differ/Diff+AppKit.swift [hunk 2]
```diff
@@ -51,11 +105,12 @@ extension NSTableView {
         )
     }
 
+    @available(*, deprecated, message: "Use `apply(_ patches: …)` based on `ExtendedPatch` instead. Deprecated because it has errors animating multiple moves.")
     public func apply(
         _ diff: ExtendedDiff,
         deletionAnimation: NSTableView.AnimationOptions = [],
         insertionAnimation: NSTableView.AnimationOptions = [],
-        indexPathTransform: (IndexPath) -> IndexPath = { $0 }
+        indexPathTransform: (IndexPath) -> IndexPath
```
Comment (jenox) [inline Sources/Differ/Diff+AppKit.swift:113]: Removing the default parameter value here is going to break clients relying on it. For the other methods, I guess calls typecheck to the new, fixed versions in case the argument is omitted.

Why are me moving away from `IndexPath`s to one-dimensional indices though? Is this a macOS/`NSTableView` thing?

Comment (hannesoid) [reply Sources/Differ/Diff+AppKit.swift:113]: My idea is:
- ideally previous callers should be migrated to the new implementation which doesn't have the solved bug, and they don't even need to know about it. I'm confident that the initial bugs with regard to moves is fixed, and nobody should need to update the call to take advantage of that. unless: 
- I left the indexPathTransform in there because for cases where there is a non-identity transform defined, it is not immediately clear how to map this to the actually only row-based approach. People who declare an indexPathTransform will then meet the deprecation warning and can transition

macOS NSTableView doesn't have sections like UITableView, and operates on `row: Int` when addressing rows. Since we only operating on rows (not on columns, which NSTableView supports),  `IndexPath`'s typical two dimensions don't translate to rows in a  meaningful way

File: Sources/Differ/Diff+AppKit.swift (hunk 1 above)
Comment (hannesoid) [reply Sources/Differ/Diff+AppKit.swift:46]: Yes I'd also prefer that, however I at first didn't see a `extendedPatch` function/init based that allowed to specify the `isEqual` to be used. Do you have suggestion what to use for this? Would `extendedDiff(…isEqual:…).patch(…)` be the right thing?
(I was a bit hesitant to introduce a new function or parameter to `extendedPatch`, as it seemed to reach further into the codebase rather than just fixing the `NSTableView` adapter)

Comment (jenox) [reply Sources/Differ/Diff+AppKit.swift:46]: I believe `extendedDiff(_:isEqual:).patch(from:to:sort:)` would be the right thing, though I'm a little confused as to why `patch(from:to:sort:)` doesn't require the comparator as well.

File: Sources/Differ/Diff+AppKit.swift (hunk 2 above)
Comment (jenox) [reply Sources/Differ/Diff+AppKit.swift:113]: Okay, let me rephrase. As far as I understood, the bug you fixed originated from using `ExtendedDiff` vs `ExtendedPatch<T>`.

We therefore cannot fix `apply(_: ExtendedDiff, ...)` and should deprecate this method as you did. However, I'd leave the default parameter value so that users relying on the default parameter actually see the deprecation instead of getting a compiler error.

In contrast, we _can_ fix the two `animateRowChanges(...)` methods without changing their signature, meaning all users get to benefit from the fix without having to change anything. If you are confident that the new behavior is correct, that's what we should aim for. However, with your current changes, only users who don't use `indexPathTransform` would get the fix.

Can we keep the `indexPathTransform` parameters, or did I miss something and they are related to the bug fix as well? Of course you could then add the `rowIndexTransform`-versions as an overlay to those, though I'd refrain from deprecating the `indexPathTransform`-based ones.

Comment (hannesoid) [reply Sources/Differ/Diff+AppKit.swift]: Ah yes got it now. You're right, on the `apply(…diff:…)` method removing the default param doesn't help.
With regard to the indexPathTransform: They are not related to the bugfix, it just confused me for a bit when I was confronted with them since they don't mean too much in the NSTableView.

**Comment by hannesoid** [comment]
@jenox please have a look at the new code. By the way, I - and probably users who's compilation won't break :) - appreciate you taking the time to look at this in such detail!

**Comment by jenox** [comment]
@hannesoid Looks good in my opinion now!

I can't merge things here though, @tonyarnold will need to do that.

File: Sources/Differ/Diff+AppKit.swift [hunk 3]
```diff
@@ -1,31 +1,86 @@
...
         oldData: T,
         newData: T,
         deletionAnimation: NSTableView.AnimationOptions = [],
         insertionAnimation: NSTableView.AnimationOptions = [],
-        indexPathTransform: (IndexPath) -> IndexPath = { $0 }
+        rowIndexTransform: (Int) -> Int = { $0 }
     ) where T.Element: Equatable {
         apply(
-            oldData.extendedDiff(newData),
+            extendedPatch(from: oldData, to: newData),
```
Comment (tonyarnold) [inline Sources/Differ/Diff+AppKit.swift:49]: Not that I'm against this change, but why the move to using a pure function? Just for cleanliness/clarity of use?

File: Sources/Differ/Diff+AppKit.swift [hunk 4]
```diff
@@ -34,23 +89,52 @@ extension NSTableView {
     ///   - isEqual:            A function comparing two elements of `T`
     ///   - deletionAnimation:  Animation type for deletions
     ///   - insertionAnimation: Animation type for insertions
-    ///   - indexPathTransform: Closure which transforms zero-based `IndexPath` to desired `IndexPath`
+    ///   - indexPathTransform: Closure which transforms zero-based `IndexPath` to desired `IndexPath`. Only the `.item` is used, not the `.section`.
+    @available(*, deprecated, message: "Use `animateRowChanges(…rowIndexTransform:)`instead. Deprecated because indexPaths are not used in `NSTableView` rows, just Integer row indices.")
```
Comment (tonyarnold) [inline Sources/Differ/Diff+AppKit.swift:93]: Is is possible to use the "rename" variant of the availability flag here so that Xcode offers a fix-it? For example: 

```swift
@available(*, deprecated, renamed: "loadData")
func fetchData() { }
```

File: Sources/Differ/Diff+AppKit.swift (hunk 2 above)
Comment (tonyarnold) [reply Sources/Differ/Diff+AppKit.swift]: I concur that the default parameter should be restored here, unless it's not required for some reason — I don't mind breaking changes when they are warranted, but this seems like unnecessary breakage to me, doubly so given that this method is deprecated.

File: Sources/Differ/Diff+AppKit.swift (hunk 3 above)
Comment (hannesoid) [reply Sources/Differ/Diff+AppKit.swift:49]: Ah, no strong opinion involved in this. I'm just new to the codebase and found this the shortest way to create the patches. Apart from this function I'm only aware of writing inline what the function does, like `oldData.extendedDiff(newData).patch(from:oldData,to:newData)`, is there another way you would recommend?

File: Sources/Differ/Diff+AppKit.swift (hunk 4 above)
Comment (hannesoid) [reply Sources/Differ/Diff+AppKit.swift:93]: Hmm, it's not just a rename, on call site the fixit is probably a great experience because then you'd have your indexPathTransform go into the rowIndexTransform param which won't compile initially. However, this way you wouldn't need to look up the method signature which is a pro.
Given this, should I try it?

Comment (hannesoid) [reply Sources/Differ/Diff+AppKit.swift:93]: ping @tonyarnold 🙂

Comment (tonyarnold) [reply Sources/Differ/Diff+AppKit.swift:93]: No, I think it's fair that this is API breakage, especially given that we're discussing the next release being breaking (version 2.0.0) in #71. I'm happy for this to be merged as-is.

File: Sources/Differ/Diff+AppKit.swift (hunk 3 above)
Comment (tonyarnold) [reply Sources/Differ/Diff+AppKit.swift:49]: No, that's completely fine - I tend to lean into methods more than free functions, so I was curious what your thinking was 👍

**Comment by tonyarnold** [comment]
I'm going to merge this into `master` now, but feel free to open new PRs to adjust anything that needs attention. I'll be away from my office for a bit this week, but I'd be keen to see work progress on #71 before cutting a new 2.0 release (just so we don't have to bump to v 3.0 due to further API breakage on the iOS side).

**Comment by hannesoid** [comment]
thanks @tonyarnold!
